test/example/SnapshotGateTest.js

```diff
@@ -0,0 +1,717 @@
+        // Upload the snapshot
+        await snapshotGate.connect(deployer).appendToSnapshot(snapshot);
+
+        // Freeze the snapshot
+        await snapshotGate.connect(deployer).freezeSnapshot();
+
+        // Grab an entry from the snapshot
+        let entry = snapshot[Math.floor(snapshot.length / 3)];
+
+        // Get the account to make the call with
+        let holder = holderByAddress[entry.owner];
+
+        // Commit to the gated offer
+        offerId = entry.tokenId; // Offer, token, and group ids are aligned for the sake of sanity,
+        await snapshotGate.connect(holder).commitToGatedOffer(entry.owner, offerId, entry.tokenId, { value: price });
+
+        // Check that the committed snapshot token is now marked used
+        const response = await snapshotGate.connect(rando).checkSnapshot(entry.tokenId, entry.owner);
+
+        // Expect owned value to match snapshot value for holder
+        expect(response.owned.toString()).to.equal(entry.amount);
+
+        // Expect used value to be one
+        expect(response.used.toString()).to.equal("1");
+      });
+
+      it("should return expected values after a commit when price is in ERC20 token", async function () {
+        // Upload the snapshot
+        await snapshotGate.connect(deployer).appendToSnapshot(snapshot);
+
+        // Freeze the snapshot
+        await snapshotGate.connect(deployer).freezeSnapshot();
+
+        // Grab an entry from the snapshot
+        let entry = snapshot[Math.floor(snapshot.length / 3)];
+
+        // ERC20 offers are in second batch
+        offerId = String(Number(entry.tokenId) + snapshotTokenCount);
+
+        // Get the account to make the call with
+        let holder = holderByAddress[entry.owner];
+
+        // Commit to the gated offer
+        await snapshotGate.connect(holder).commitToGatedOffer(entry.owner, offerId, entry.tokenId);
+
+        // Check that the committed snapshot token is now marked used
+        const response = await snapshotGate.connect(rando).checkSnapshot(entry.tokenId, entry.owner);
+
+        // Expect owned value to match snapshot value for holder
+        expect(response.owned.toString()).to.equal(entry.amount);
+
+        // Expect used value to be one
+        expect(response.used.toString()).to.equal("1");
+      });
+    });
+
+    context("👉 ownerOf()", async function () {
+      it("should report gate as owner of all tokens when tx not in-flight", async function () {
+        // Upload the snapshot, creating all tokens in the process
+        await snapshotGate.connect(deployer).appendToSnapshot(snapshot);
+
+        // Expect the owner of all the tokens to be the gate contract itself
+        for (let i = 1; i <= snapshotTokenCount; i++) {
+          const tokenId = i.toString();
+          const owner = await snapshotGate.connect(rando).ownerOf(tokenId);
+          expect(owner).to.equal(snapshotGate.address);
+        }
+      });
+    });
+  });
+});
```

Comment:
Need test case to check that buyer is reported as owner when transaction IS in-flight

Reply:
This would require building and testing a mock protocol contract that calls back in. I think that's outside the scope of the example,. We can infer that it works correctly because positive unit tests are already showing that the protocol allows commit to offer for these gated offers, which can only happen if the buyer is reported as being the owner of the token when the protocol calls back in the commitToOffer transaction flow.

Reply:
I don't understand what you mean. This test file deploys several protocol facets in the main `beforEach`. Most of our test files do this. No mock protocol is needed. The real one is available.